OMCompiler/SimulationRuntime/fmi/export/openmodelica/fmu2_model_interface.c.inc

```diff
@@ -1153,19 +1153,190 @@ fmi2Status fmi2SetString(fmi2Component c, const fmi2ValueReference vr[], size_t
+    // infoStreamPrint(LOG_STDOUT, 0, "Copy Integer parameter %s = %ld", fmudata->modelData->integerParameterData[i].info.name, internal_state->integerParameters[i]);
+  }
+
+  // copy Boolean parameter variables
+  internal_state->booleanParameter = (modelica_boolean*)calloc(fmudata->modelData->nParametersBoolean, sizeof(modelica_boolean));
+  for (int i = 0; i < fmudata->modelData->nParametersBoolean; ++i)
+  {
+    internal_state->booleanParameter[i] = fmudata->modelData->booleanParameterData[i].attribute.start;
+    //infoStreamPrint(LOG_STDOUT, 0, "copy Boolean parameter %s = %s", fmudata->modelData->booleanParameterData[i].info.name, internal_state->booleanParameters[i] ? "true" : "false");
+  }
+
+  // copy String parameter variables
+  internal_state->stringParameter = (modelica_string*) omc_alloc_interface.malloc_uncollectable(fmudata->modelData->nParametersString * sizeof(modelica_string));
+  for (int i = 0; i < fmudata->modelData->nParametersString; ++i)
+  {
+    internal_state->stringParameter[i] = fmudata->modelData->stringParameterData[i].attribute.start;
+    //infoStreamPrint(LOG_STDOUT, 0, "copy String parameter %s = %s", fmudata->modelData->stringParameterData[i].info.name, MMC_STRINGDATA(internal_state->stringParameters[i]));
+  }
+
+  //infoRingBuffer( fmudata->simulationData);
+  //printRingBufferSimulationData(fmudata->simulationData, fmudata); // original ringBuffer data
+  //printRingBufferSimulationData(internal_state->simulationData, fmudata); // copied ringBuffer data
+
+  // return the fmu state
+  *FMUstate = (fmi2FMUstate) internal_state;
+  return fmi2OK;
 }
 
 fmi2Status fmi2SetFMUstate(fmi2Component c, fmi2FMUstate FMUstate)
 {
-  return unsupportedFunction(c, "fmi2SetFMUstate", modelInstantiated|modelInitializationMode|modelEventMode|modelContinuousTimeMode|modelTerminated|modelError);
+  ModelInstance *comp = (ModelInstance *) c;
+  int meStates = modelInstantiated|modelInitializationMode|modelEventMode;
+  int csStates = modelInstantiated|modelInitializationMode|modelEventMode|modelContinuousTimeMode;
+  if (invalidState(comp, "fmi2GetFMUstate", meStates, csStates))
+    return fmi2Error;
+
+  INTERNAL_FMU_STATE * internal_state = (INTERNAL_FMU_STATE *) FMUstate;
+  DATA* fmudata = (DATA *) comp->fmuData;
+
+  //printRingBufferSimulationData(internal_state->simulationData, fmudata); // copied ringBuffer data
+
+  // override the SIMULATION_DATA with INTERNAL_FMU_STATE
+  for (int i = 0; i < ringBufferLength(internal_state->simulationData); i++)
+  {
+    SIMULATION_DATA *sdata = (SIMULATION_DATA *)getRingData(internal_state->simulationData, i);
+    fmudata->localData[i]->timeValue = sdata->timeValue;
+    fmudata->localData[i]->realVars = sdata->realVars;
+    fmudata->localData[i]->integerVars = sdata->integerVars;
+    fmudata->localData[i]->booleanVars = sdata->booleanVars;
+    fmudata->localData[i]->stringVars = sdata->stringVars;
```

Comment:
Using memcpy here and also in the later statements?

Reply:
@this is fixed with this commit https://github.com/OpenModelica/OpenModelica/pull/8638/commits/099afca47fd4233e485356c16cbf416a03a2cf1c